### PR TITLE
fetch photo bytes from storage path and retrieve rubric criteria via …

### PR DESCRIPTION
**Pull request by jpozuna**
# What
scoring.py — scorer now accepts photo_path: Optional[str] instead of UploadFile; downloads bytes from Supabase Storage at scoring time, safe for background task execution
scoring.py — full RAG pipeline added: submission text is embedded via text-embedding-3-small, cosine similarity run against task_criteria embeddings, top 5 matches injected into Claude prompt alongside full criteria list
scoring.py — _mark_submission_error() helper centralizes error writes; granular try/catch wraps storage download, GPT-4o describe, and embedding call independently so each failure mode is traceable
scoring.py — retrieved_criteria persisted in ai_result on all terminal paths (approved, flagged, invalid JSON) for auditability
submissions.py — photo uploaded to Storage during request lifecycle, path stored in photo_url; upload failure is non-fatal, submission still created and scorer marks it error

# Why
UploadFile is closed before background tasks run — fetching by path makes scoring decoupled from the request lifecycle and safe to retry. RAG retrieval grounds Claude's scores in organizer-defined rubric criteria

### CODE REVIEW

File: backend/services/scoring.py [hunk 1]
```diff
@@ -83,30 +131,53 @@ async def score_submission(submission_id: str, task_id: str, team_id: str, text_
...
 
         # Get GPT-4o description if photo submission
         gpt4o_description = None
-        if photo:
-            photo_bytes = await photo.read()
-            import base64
+        if photo_path:
+            try:
+                photo_bytes = supabase.storage.from_(_storage_bucket()).download(photo_path)
+            except Exception as e:
```
[the remark this answers is not in the document]
Comment (jpozuna) [reply backend/services/scoring.py:155]: I now wrapped the download in anyio.to_thread.run_sync so the async background task doesn’t block the event loop.

File: backend/services/scoring.py [hunk 2]
```diff
@@ -83,30 +131,53 @@ async def score_submission(submission_id: str, task_id: str, team_id: str, text_
...
+            try:
+                response = openai_client.chat.completions.create(
+                    model="gpt-4o",
+                    messages=[
+                        {
+                            "role": "user",
+                            "content": [
+                                {"type": "image_url", "image_url": {"url": f"data:image/jpeg;base64,{b64}"}},
+                                {"type": "text", "text": "Describe exactly what you see in this image in detail."},
+                            ],
```
[the remark this answers is not in the document]
Comment (jpozuna) [reply backend/services/scoring.py:175]: Now I infer the MIME type from photo_path instead of always image/jpeg.

File: backend/routes/submissions.py [hunk 3]
```diff
@@ -46,15 +55,34 @@ async def create_submission(
...
+        photo_bytes = await photo.read()
+        content_type = (photo.content_type or "application/octet-stream").strip()
+        ext = (content_type.split("/")[-1] if "/" in content_type else "bin") or "bin"
+        photo_path = f"submissions/{submission_id}.{ext}"
+        try:
+            supabase.storage.from_(_storage_bucket()).upload(
+                photo_path,
+                photo_bytes,
+                file_options={"content-type": content_type, "upsert": True},
+            )
```
[the remark this answers is not in the document]
Comment (jpozuna) [reply backend/routes/submissions.py]: Now I offloaded the upload to a worker thread with anyio.to_thread.run_sync to keep the request handler non-blocking.

File: backend/routes/submissions.py [hunk 4]
```diff
@@ -46,15 +55,34 @@ async def create_submission(
...
+        except Exception:
+            # If photo upload fails, still create the submission and let scoring mark it as error.
+            photo_path = None
```
[the remark this answers is not in the document]
Comment (jpozuna) [reply backend/routes/submissions.py]: Fixed. If upload fails I create the submission with status="error". Also scorer returns error if theres no content.

File: backend/services/scoring.py [hunk 5]
```diff
@@ -65,11 +67,57 @@ def _parse_score_json(text: str, *, max_points: int) -> ScoreResult:
 
     return ScoreResult(score=score, confidence=confidence, rationale=rationale, raw=obj)
 
-async def score_submission(submission_id: str, task_id: str, team_id: str, text_answer: str, photo):
+
+def _storage_bucket() -> str:
+    return os.getenv("SUPABASE_STORAGE_BUCKET", "submissions").strip() or "submissions"
+
+
```
[the remark this answers is not in the document]
Comment (jpozuna) [reply backend/services/scoring.py]: Addressed by moving bucket selection to a shared helper and importing it.

File: backend/services/scoring.py [hunk 6]
```diff
@@ -134,6 +205,50 @@ async def score_submission(submission_id: str, task_id: str, team_id: str, text_
...
+                if sim == float("-inf"):
+                    continue
+                scored.append((sim, row))
+
+            scored.sort(key=lambda t: t[0], reverse=True)
+            for sim, row in scored[:5]:
+                retrieved_criteria.append(
+                    {"value": row.get("value"), "criteria_type": row.get("criteria_type"), "similarity": sim}
+                )
+
```
[the remark this answers is not in the document]
Comment (jpozuna) [reply backend/services/scoring.py:270]: Added unit tests for _cosine_similarity edge cases (empty vectors, mismatched lengths, zero norms) to cover the riskiest part of the RAG retrieval math.
